darts/timeseries.py

```diff
@@ -1807,18 +1807,21 @@ def map(self,
             fn: Union[Callable[[np.number], np.number],
                       Callable[[Union[pd.Timestamp, int], np.number], np.number]]) -> 'TimeSeries':  # noqa: E501
         """
-        Applies the function `fn` elementwise to all values in this TimeSeries.
-        Returns a new TimeSeries instance. If `fn` takes 1 argument it is simply applied elementwise.
-        If it takes 2 arguments, it is applied elementwise on the (timestamp, value) tuples.
-
-        At the moment this function works only on deterministic time series (i.e., made of 1 sample).
+        Applies the function `fn` to the underlying NumPy array containing this series' values.
+        Returns a new TimeSeries instance. If `fn` takes 1 argument it is simply applied on the backing array
+        of shape (time, n_components, n_samples).
+        If it takes 2 arguments, it is applied repeteadly on the (ts, value[ts]) tuples, where
+        "ts" denotes a timestamp value, and "value[ts]" denote the array of values at this timestamp, of shape
+        (n_components, n_samples).
 
         Parameters
         ----------
         fn
-            Either a function which takes a value and returns a value ie. `f(x) = y`
-            Or a function which takes a value and its timestamp and returns a value ie. `f(timestamp, x) = y`
-            The type of `timestamp` is either `pd.Timestamp` (if the series is indexed with a DatetimeIndex),
+            Either a function which takes a NumPy array and returns a NumPy array of same shape; 
+            e.g., `lambda x: x ** 2`, `lambda x: x / x.shape[0]` or `np.log`.
+            It can also be a function which takes a timestamp and array, and returns a new array of same shape;
+            e.g., `lambda ts, x: x / ts.days_in_month`.
+            The type of `ts` is either `pd.Timestamp` (if the series is indexed with a DatetimeIndex),
```

Comment:
Why again do we expect to have a `pd.Timestamp` or `int` instead of a `pd.DatetimeIndex` or `(pd.Int64Index, pd.RangeIndex)`?

`lambda ts, x: x / ts.days_in_month` can also be applied on the entire`time_index`. ->
`lambda time_index, x: x / time_index.days_in_month`

Similar to darts.utils.data.encoders.CallableIndexEncoder


Could this avoid having to iterate over every time stamp anymore as you mentioned below?

Maybe I'm missing something :)

Reply:
You are right it would be nicer implementation-wise, however the signature of the resulting function is not as nice.
For instance instead of writing
```
lambda ts, vals: vals/ts.days_in_month
```

We would have to write:
```
lambda idx, vals: (vals.T/idx.days_in_month.values).T
```

More efficient for sure, but not that nice, no?

Reply:
I see what you mean. What do you think about finding the shortest path to compute the result? Something like this

```
            # go over shortest amount of iterations, either over time steps or components and samples
            if self.n_timesteps <= self.n_components * self.n_samples:
                new_vals = np.vstack([
                    np.expand_dims(fn(self.time_index[i], self._xa[i, :, :]), axis=0) for i in range(self.n_timesteps)
                ])
            else:
                new_vals = np.stack([
                    np.column_stack([fn(self.time_index, self._xa[:, i, j]) for j in range(self.n_samples)])
                    for i in range(self.n_components)
                ], axis=1)
            new_xa.values = new_vals
```
this could reduce cost by quite a bit (see below). I also assume that in most cases during training `n_timesteps >> n_components * n_samples` and for prediction `n_timestamps << n_components * n_samples`. This approach could leverage the fast route for both. WDYT?

![image](https://user-images.githubusercontent.com/84317528/145776437-bf13fa04-fe9e-45bd-8d66-3f61da25b1f4.png)

Reply:
OK nice, I like it :) thanks!